diff --git a/src/cmd/compile/internal/arm64/ssa.go b/src/cmd/compile/internal/arm64/ssa.go
index 753cef743b33cc719e8b84c9f5a9b8992e709c4b..fdda425900b21de1e914ff18dad144b25b9fda50 100644
--- a/src/cmd/compile/internal/arm64/ssa.go
+++ b/src/cmd/compile/internal/arm64/ssa.go
@@ -1121,6 +1121,11 @@ func ssaGenValue(s *ssagen.State, v *ssa.Value) {
 		p.To.Name = obj.NAME_EXTERN
 		// AuxInt encodes how many buffer entries we need.
 		p.To.Sym = ir.Syms.GCWriteBarrier[v.AuxInt-1]
+	case ssa.OpARM64LoweredMemEq:
+		p := s.Prog(obj.ACALL)
+		p.To.Type = obj.TYPE_MEM
+		p.To.Name = obj.NAME_EXTERN
+		p.To.Sym = ir.Syms.Memequal
 
 	case ssa.OpARM64LoweredPanicBoundsA, ssa.OpARM64LoweredPanicBoundsB, ssa.OpARM64LoweredPanicBoundsC:
 		p := s.Prog(obj.ACALL)
diff --git a/src/cmd/compile/internal/ir/symtab.go b/src/cmd/compile/internal/ir/symtab.go
index e2da710f025878c12b602e1e23945f4b8c501e11..61c8c0c8bd2a2c633fabe84fba61a9e7374b0650 100644
--- a/src/cmd/compile/internal/ir/symtab.go
+++ b/src/cmd/compile/internal/ir/symtab.go
@@ -32,6 +32,7 @@ type symsStruct struct {
 	InterfaceSwitch   *obj.LSym
 	MallocGC          *obj.LSym
 	Memmove           *obj.LSym
+	Memequal          *obj.LSym
 	Msanread          *obj.LSym
 	Msanwrite         *obj.LSym
 	Msanmove          *obj.LSym
diff --git a/src/cmd/compile/internal/ssa/_gen/ARM64.rules b/src/cmd/compile/internal/ssa/_gen/ARM64.rules
index 01fe3a74f7d551b25afc870ad858755a766dfd05..e73f98b4d4faa036970b45029d9f714ea620e858 100644
--- a/src/cmd/compile/internal/ssa/_gen/ARM64.rules
+++ b/src/cmd/compile/internal/ssa/_gen/ARM64.rules
@@ -532,6 +532,9 @@
 		(ADDconst <src.Type> src [s-16])
 		mem)
 
+// mem ops
+(MemEq ...) => (LoweredMemEq ...)
+
 // calls
 (StaticCall  ...) => (CALLstatic  ...)
 (ClosureCall ...) => (CALLclosure ...)
diff --git a/src/cmd/compile/internal/ssa/_gen/ARM64Ops.go b/src/cmd/compile/internal/ssa/_gen/ARM64Ops.go
index ebb7ed52997c43e46234338f3d1a07454c990dab..c333c89ddc44d2d2b0c5abab0f0faa976a0055f6 100644
--- a/src/cmd/compile/internal/ssa/_gen/ARM64Ops.go
+++ b/src/cmd/compile/internal/ssa/_gen/ARM64Ops.go
@@ -517,6 +517,7 @@ func init() {
 
 		// pseudo-ops
 		{name: "LoweredNilCheck", argLength: 2, reg: regInfo{inputs: []regMask{gpg}}, nilCheck: true, faultOnNilArg0: true}, // panic if arg0 is nil.  arg1=mem.
+		{name: "LoweredMemEq", argLength: 4, reg: regInfo{inputs: []regMask{r0, r1, r2}, outputs: []regMask{r0}, clobbers: buildReg("R0 R1 R2 R3 R4 R5 R6 R7 R9 R30 F0 F1 F2 F3 F4 F5 F6 F7 F8 F9 F10 F11")}, typ: "Bool", faultOnNilArg0: true, faultOnNilArg1: true, clobberFlags: true, call: true}, // arg0, arg1 - pointers to memory, arg2=size, arg3=mem.
 
 		{name: "Equal", argLength: 1, reg: readflags},            // bool, true flags encode x==y false otherwise.
 		{name: "NotEqual", argLength: 1, reg: readflags},         // bool, true flags encode x!=y false otherwise.
diff --git a/src/cmd/compile/internal/ssa/_gen/generic.rules b/src/cmd/compile/internal/ssa/_gen/generic.rules
index b178a1add6de29eb2181106a311dd7ac0fd0a099..09384df7bbbd41b21f5bc8340c662d93972c98f1 100644
--- a/src/cmd/compile/internal/ssa/_gen/generic.rules
+++ b/src/cmd/compile/internal/ssa/_gen/generic.rules
@@ -2152,6 +2152,48 @@
   && isSamePtr(p, q)
   => (MakeResult (ConstBool <typ.Bool> [true]) mem)
 
+(MemEq sptr (Addr {scon} (SB)) (Const64 [1]) mem)
+  && symIsRO(scon)
+  => (Eq8 (Load <typ.Int8> sptr mem) (Const8 <typ.Int8> [int8(read8(scon,0))]))
+
+(MemEq (Addr {scon} (SB)) sptr (Const64 [1]) mem)
+  && symIsRO(scon)
+  => (Eq8 (Load <typ.Int8> sptr mem) (Const8 <typ.Int8> [int8(read8(scon,0))]))
+
+(MemEq sptr (Addr {scon} (SB)) (Const64 [2]) mem)
+  && symIsRO(scon)
+  && canLoadUnaligned(config)
+  => (Eq16 (Load <typ.Int16> sptr mem) (Const16 <typ.Int16> [int16(read16(scon,0,config.ctxt.Arch.ByteOrder))]))
+
+(MemEq (Addr {scon} (SB)) sptr (Const64 [2]) mem)
+  && symIsRO(scon)
+  && canLoadUnaligned(config)
+  => (Eq16 (Load <typ.Int16> sptr mem) (Const16 <typ.Int16> [int16(read16(scon,0,config.ctxt.Arch.ByteOrder))]))
+
+(MemEq sptr (Addr {scon} (SB)) (Const64 [4]) mem)
+  && symIsRO(scon)
+  && canLoadUnaligned(config)
+  => (Eq32 (Load <typ.Int32> sptr mem) (Const32 <typ.Int32> [int32(read32(scon,0,config.ctxt.Arch.ByteOrder))]))
+
+(MemEq (Addr {scon} (SB)) sptr (Const64 [4]) mem)
+  && symIsRO(scon)
+  && canLoadUnaligned(config)
+  => (Eq32 (Load <typ.Int32> sptr mem) (Const32 <typ.Int32> [int32(read32(scon,0,config.ctxt.Arch.ByteOrder))]))
+
+(MemEq sptr (Addr {scon} (SB)) (Const64 [8]) mem)
+  && symIsRO(scon)
+  && canLoadUnaligned(config) && config.PtrSize == 8
+  => (Eq64 (Load <typ.Int64> sptr mem) (Const64 <typ.Int64> [int64(read64(scon,0,config.ctxt.Arch.ByteOrder))]))
+
+(MemEq (Addr {scon} (SB)) sptr (Const64 [8]) mem)
+  && symIsRO(scon)
+  && canLoadUnaligned(config) && config.PtrSize == 8
+  => (Eq64 (Load <typ.Int64> sptr mem) (Const64 <typ.Int64> [int64(read64(scon,0,config.ctxt.Arch.ByteOrder))]))
+
+(MemEq _ _ (Const64 [0]) _) => (ConstBool <typ.Bool> [true])
+
+(MemEq p q _ _) && isSamePtr(p, q) => (ConstBool <typ.Bool> [true])
+
 // Turn known-size calls to memclrNoHeapPointers into a Zero.
 // Note that we are using types.Types[types.TUINT8] instead of sptr.Type.Elem() - see issue 55122 and CL 431496 for more details.
 (SelectN [0] call:(StaticCall {sym} sptr (Const(64|32) [c]) mem))
diff --git a/src/cmd/compile/internal/ssa/_gen/genericOps.go b/src/cmd/compile/internal/ssa/_gen/genericOps.go
index 1f6ad4e16d98f8803ed54473ac360db487bb2b48..3eb65166af64efc29d7aea1042437ecf7080ec9e 100644
--- a/src/cmd/compile/internal/ssa/_gen/genericOps.go
+++ b/src/cmd/compile/internal/ssa/_gen/genericOps.go
@@ -662,6 +662,9 @@ var genericOps = []opData{
 	// Prefetch instruction
 	{name: "PrefetchCache", argLength: 2, hasSideEffects: true},         // Do prefetch arg0 to cache. arg0=addr, arg1=memory.
 	{name: "PrefetchCacheStreamed", argLength: 2, hasSideEffects: true}, // Do non-temporal or streamed prefetch arg0 to cache. arg0=addr, arg1=memory.
+
+	// Helper instruction which is equivalent to runtime.memequal, but can be sometimes lowered more efficiently.
+	{name: "MemEq", argLength: 4, typ: "Bool"}, // arg0=ptr0, arg1=ptr1, arg2=size, arg3=memory.
 }
 
 //     kind          controls          successors   implicit exit

diff --git a/src/cmd/compile/internal/ssagen/intrinsics.go b/src/cmd/compile/internal/ssagen/intrinsics.go
index 6b58e7e5914538de18baaf04a507bd925147f888..6a8ef4320f52295ffde73ee637396faaef03f894 100644
--- a/src/cmd/compile/internal/ssagen/intrinsics.go
+++ b/src/cmd/compile/internal/ssagen/intrinsics.go
@@ -195,6 +195,14 @@ func initIntrinsics(cfg *intrinsicBuildConfig) {
 		},
 		sys.AMD64, sys.I386, sys.ARM64, sys.ARM, sys.Loong64, sys.S390X)
 
+	if base.Flag.MemOp != 0 {
+		addF("runtime", "memequal",
+			func(s *state, n *ir.CallExpr, args []*ssa.Value) *ssa.Value {
+				return s.newValue4(ssa.OpMemEq, s.f.Config.Types.Bool, args[0], args[1], args[2], s.mem())
+			},
+			sys.ARM64)
+	}
+
 	if cfg.goppc64 >= 10 {
 		// Use only on Power10 as the new byte reverse instructions that Power10 provide
 		// make it worthwhile as an intrinsic
diff --git a/src/cmd/compile/internal/ssagen/ssa.go b/src/cmd/compile/internal/ssagen/ssa.go
index e241e9b9bc0f4f86c64d98ac759dc69390f1efdf..82db4b93648036cb4eb6ccaca460c534f1e514a4 100644
--- a/src/cmd/compile/internal/ssagen/ssa.go
+++ b/src/cmd/compile/internal/ssagen/ssa.go
@@ -130,6 +130,7 @@ func InitConfig() {
 	ir.Syms.InterfaceSwitch = typecheck.LookupRuntimeFunc("interfaceSwitch")
 	ir.Syms.MallocGC = typecheck.LookupRuntimeFunc("mallocgc")
 	ir.Syms.Memmove = typecheck.LookupRuntimeFunc("memmove")
+	ir.Syms.Memequal = typecheck.LookupRuntimeFunc("memequal")
 	ir.Syms.Msanread = typecheck.LookupRuntimeFunc("msanread")
 	ir.Syms.Msanwrite = typecheck.LookupRuntimeFunc("msanwrite")
 	ir.Syms.Msanmove = typecheck.LookupRuntimeFunc("msanmove")
diff --git a/src/cmd/compile/internal/ssa/_gen/generic.rules b/src/cmd/compile/internal/ssa/_gen/generic.rules
index 09384df7bbbd41b21f5bc8340c662d93972c98f1..1f565f16c02ffef45b9ca1d3221517eac54d4a13 100644
--- a/src/cmd/compile/internal/ssa/_gen/generic.rules
+++ b/src/cmd/compile/internal/ssa/_gen/generic.rules
@@ -2152,43 +2152,36 @@
   && isSamePtr(p, q)
   => (MakeResult (ConstBool <typ.Bool> [true]) mem)
 
-(MemEq sptr (Addr {scon} (SB)) (Const64 [1]) mem)
-  && symIsRO(scon)
-  => (Eq8 (Load <typ.Int8> sptr mem) (Const8 <typ.Int8> [int8(read8(scon,0))]))
+(MemEq sptr tptr (Const64 [1]) mem)
+  => (Eq8 (Load <typ.Int8> sptr mem) (Load <typ.Int8> tptr mem))
 
-(MemEq (Addr {scon} (SB)) sptr (Const64 [1]) mem)
+(Load <typ.Int8> sptr:(Addr {scon} (SB)) mem)
   && symIsRO(scon)
-  => (Eq8 (Load <typ.Int8> sptr mem) (Const8 <typ.Int8> [int8(read8(scon,0))]))
+  => (Const8 <typ.Int8> [int8(read8(scon,0))])
 
-(MemEq sptr (Addr {scon} (SB)) (Const64 [2]) mem)
-  && symIsRO(scon)
+(MemEq sptr tptr (Const64 [2]) mem)
   && canLoadUnaligned(config)
-  => (Eq16 (Load <typ.Int16> sptr mem) (Const16 <typ.Int16> [int16(read16(scon,0,config.ctxt.Arch.ByteOrder))]))
+  => (Eq16 (Load <typ.Int16> sptr mem) (Load <typ.Int16> tptr mem))
 
-(MemEq (Addr {scon} (SB)) sptr (Const64 [2]) mem)
+(Load <typ.Int16> sptr:(Addr {scon} (SB)) mem)
   && symIsRO(scon)
-  && canLoadUnaligned(config)
-  => (Eq16 (Load <typ.Int16> sptr mem) (Const16 <typ.Int16> [int16(read16(scon,0,config.ctxt.Arch.ByteOrder))]))
+  => (Const16 <typ.Int16> [int16(read16(scon,0,config.ctxt.Arch.ByteOrder))])
 
-(MemEq sptr (Addr {scon} (SB)) (Const64 [4]) mem)
-  && symIsRO(scon)
+(MemEq sptr tptr (Const64 [4]) mem)
   && canLoadUnaligned(config)
-  => (Eq32 (Load <typ.Int32> sptr mem) (Const32 <typ.Int32> [int32(read32(scon,0,config.ctxt.Arch.ByteOrder))]))
+  => (Eq32 (Load <typ.Int32> sptr mem) (Load <typ.Int32> tptr mem))
 
-(MemEq (Addr {scon} (SB)) sptr (Const64 [4]) mem)
+(Load <typ.Int32> sptr:(Addr {scon} (SB)) mem)
   && symIsRO(scon)
-  && canLoadUnaligned(config)
-  => (Eq32 (Load <typ.Int32> sptr mem) (Const32 <typ.Int32> [int32(read32(scon,0,config.ctxt.Arch.ByteOrder))]))
+  => (Const32 <typ.Int32> [int32(read32(scon,0,config.ctxt.Arch.ByteOrder))])
 
-(MemEq sptr (Addr {scon} (SB)) (Const64 [8]) mem)
-  && symIsRO(scon)
+(MemEq sptr tptr (Const64 [8]) mem)
   && canLoadUnaligned(config) && config.PtrSize == 8
-  => (Eq64 (Load <typ.Int64> sptr mem) (Const64 <typ.Int64> [int64(read64(scon,0,config.ctxt.Arch.ByteOrder))]))
+  => (Eq64 (Load <typ.Int64> sptr mem) (Load <typ.Int64> tptr mem))
 
-(MemEq (Addr {scon} (SB)) sptr (Const64 [8]) mem)
+(Load <typ.Int64> sptr:(Addr {scon} (SB)) mem)
   && symIsRO(scon)
-  && canLoadUnaligned(config) && config.PtrSize == 8
-  => (Eq64 (Load <typ.Int64> sptr mem) (Const64 <typ.Int64> [int64(read64(scon,0,config.ctxt.Arch.ByteOrder))]))
+  => (Const64 <typ.Int64> [int64(read64(scon,0,config.ctxt.Arch.ByteOrder))])
 
 (MemEq _ _ (Const64 [0]) _) => (ConstBool <typ.Bool> [true])
 







diff --git a/src/cmd/compile/internal/ssa/_gen/ARM64Ops.go b/src/cmd/compile/internal/ssa/_gen/ARM64Ops.go
index c333c89ddc44d2d2b0c5abab0f0faa976a0055f6..cb60efae1eeb7e43e9108dcd68b2fb65039f0331 100644
--- a/src/cmd/compile/internal/ssa/_gen/ARM64Ops.go
+++ b/src/cmd/compile/internal/ssa/_gen/ARM64Ops.go
@@ -517,7 +517,7 @@ func init() {
 
 		// pseudo-ops
 		{name: "LoweredNilCheck", argLength: 2, reg: regInfo{inputs: []regMask{gpg}}, nilCheck: true, faultOnNilArg0: true}, // panic if arg0 is nil.  arg1=mem.
-		{name: "LoweredMemEq", argLength: 4, reg: regInfo{inputs: []regMask{r0, r1, r2}, outputs: []regMask{r0}, clobbers: buildReg("R0 R1 R2 R3 R4 R5 R6 R7 R9 R30 F0 F1 F2 F3 F4 F5 F6 F7 F8 F9 F10 F11")}, typ: "Bool", faultOnNilArg0: true, faultOnNilArg1: true, clobberFlags: true, call: true}, // arg0, arg1 - pointers to memory, arg2=size, arg3=mem.
+		{name: "LoweredMemEq", argLength: 4, reg: regInfo{inputs: []regMask{r0, r1, r2}, outputs: []regMask{r0}, clobbers: buildReg("R0 R1 R2 R16 R17 R19 R20 R21 R22 R23 R24 R30 F0 F1 F2 F3 F4 F5 F6 F7 F8 F9 F10 F11")}, typ: "Bool", faultOnNilArg0: true, faultOnNilArg1: true, clobberFlags: true}, // arg0, arg1 - pointers to memory, arg2=size, arg3=mem.
 
 		{name: "Equal", argLength: 1, reg: readflags},            // bool, true flags encode x==y false otherwise.
 		{name: "NotEqual", argLength: 1, reg: readflags},         // bool, true flags encode x!=y false otherwise.
diff --git a/src/cmd/compile/internal/ssa/opGen.go b/src/cmd/compile/internal/ssa/opGen.go
index 7383191b7d7a6d83409f7257cfbc88f031d5a444..b1879c033f5bc88e2095bed93c2e65fa735a23e0 100644
--- a/src/cmd/compile/internal/ssa/opGen.go
+++ b/src/cmd/compile/internal/ssa/opGen.go
@@ -22861,7 +22861,6 @@ var opcodeTable = [...]opInfo{
 		name:           "LoweredMemEq",
 		argLen:         4,
 		clobberFlags:   true,
-		call:           true,
 		faultOnNilArg0: true,
 		faultOnNilArg1: true,
 		reg: regInfo{
@@ -22870,7 +22869,7 @@ var opcodeTable = [...]opInfo{
 				{1, 2}, // R1
 				{2, 4}, // R2
 			},
-			clobbers: 8794213974783, // R0 R1 R2 R3 R4 R5 R6 R7 R9 R30 F0 F1 F2 F3 F4 F5 F6 F7 F8 F9 F10 F11
+			clobbers: 8794230685703, // R0 R1 R2 R16 R17 R19 R20 R21 R22 R23 R24 R30 F0 F1 F2 F3 F4 F5 F6 F7 F8 F9 F10 F11
 			outputs: []outputInfo{
 				{0, 1}, // R0
 			},
diff --git a/src/internal/bytealg/equal_arm64.s b/src/internal/bytealg/equal_arm64.s
index 408ab374e629943d5acfb5423fa46be73cf258bb..34832112619d54f9cf221b6e5524d3b768667ba1 100644
--- a/src/internal/bytealg/equal_arm64.s
+++ b/src/internal/bytealg/equal_arm64.s
@@ -32,10 +32,10 @@ TEXT runtime·memequal<ABIInternal>(SB),NOSPLIT|NOFRAME,$0-25
 	CMP	$16, R2
 	// handle specially if length < 16
 	BLO	tail
-	BIC	$0x3f, R2, R3
-	CBZ	R3, chunk16
+	BIC	$0x3f, R2, R19
+	CBZ	R19, chunk16
 	// work with 64-byte chunks
-	ADD	R3, R0, R6	// end of chunks
+	ADD	R19, R0, R24	// end of chunks
 chunk64_loop:
 	VLD1.P	(R0), [V0.D2, V1.D2, V2.D2, V3.D2]
 	VLD1.P	(R1), [V4.D2, V5.D2, V6.D2, V7.D2]
@@ -46,69 +46,69 @@ chunk64_loop:
 	VAND	V8.B16, V9.B16, V8.B16
 	VAND	V8.B16, V10.B16, V8.B16
 	VAND	V8.B16, V11.B16, V8.B16
-	CMP	R0, R6
-	VMOV	V8.D[0], R4
-	VMOV	V8.D[1], R5
-	CBZ	R4, not_equal
-	CBZ	R5, not_equal
+	CMP	R0, R24
+	VMOV	V8.D[0], R20
+	VMOV	V8.D[1], R21
+	CBZ	R20, not_equal
+	CBZ	R21, not_equal
 	BNE	chunk64_loop
 	AND	$0x3f, R2, R2
 	CBZ	R2, equal
 chunk16:
 	// work with 16-byte chunks
-	BIC	$0xf, R2, R3
-	CBZ	R3, tail
-	ADD	R3, R0, R6	// end of chunks
+	BIC	$0xf, R2, R19
+	CBZ	R19, tail
+	ADD	R19, R0, R24	// end of chunks
 chunk16_loop:
-	LDP.P	16(R0), (R4, R5)
-	LDP.P	16(R1), (R7, R9)
-	EOR	R4, R7
-	CBNZ	R7, not_equal
-	EOR	R5, R9
-	CBNZ	R9, not_equal
-	CMP	R0, R6
+	LDP.P	16(R0), (R20, R21)
+	LDP.P	16(R1), (R22, R23)
+	EOR	R20, R22
+	CBNZ	R22, not_equal
+	EOR	R21, R23
+	CBNZ	R23, not_equal
+	CMP	R0, R24
 	BNE	chunk16_loop
 	AND	$0xf, R2, R2
 	CBZ	R2, equal
 tail:
 	// special compare of tail with length < 16
 	TBZ	$3, R2, lt_8
-	MOVD	(R0), R4
-	MOVD	(R1), R5
-	EOR	R4, R5
-	CBNZ	R5, not_equal
-	SUB	$8, R2, R6	// offset of the last 8 bytes
-	MOVD	(R0)(R6), R4
-	MOVD	(R1)(R6), R5
-	EOR	R4, R5
-	CBNZ	R5, not_equal
+	MOVD	(R0), R20
+	MOVD	(R1), R21
+	EOR	R20, R21
+	CBNZ	R21, not_equal
+	SUB	$8, R2, R24	// offset of the last 8 bytes
+	MOVD	(R0)(R24), R20
+	MOVD	(R1)(R24), R21
+	EOR	R20, R21
+	CBNZ	R21, not_equal
 	B	equal
 	PCALIGN	$16
 lt_8:
 	TBZ	$2, R2, lt_4
-	MOVWU	(R0), R4
-	MOVWU	(R1), R5
-	EOR	R4, R5
-	CBNZ	R5, not_equal
-	SUB	$4, R2, R6	// offset of the last 4 bytes
-	MOVWU	(R0)(R6), R4
-	MOVWU	(R1)(R6), R5
-	EOR	R4, R5
-	CBNZ	R5, not_equal
+	MOVWU	(R0), R20
+	MOVWU	(R1), R21
+	EOR	R20, R21
+	CBNZ	R21, not_equal
+	SUB	$4, R2, R24	// offset of the last 4 bytes
+	MOVWU	(R0)(R24), R20
+	MOVWU	(R1)(R24), R21
+	EOR	R20, R21
+	CBNZ	R21, not_equal
 	B	equal
 	PCALIGN	$16
 lt_4:
 	TBZ	$1, R2, lt_2
-	MOVHU.P	2(R0), R4
-	MOVHU.P	2(R1), R5
-	CMP	R4, R5
+	MOVHU.P	2(R0), R20
+	MOVHU.P	2(R1), R21
+	CMP	R20, R21
 	BNE	not_equal
 lt_2:
 	TBZ	$0, R2, equal
 one:
-	MOVBU	(R0), R4
-	MOVBU	(R1), R5
-	CMP	R4, R5
+	MOVBU	(R0), R20
+	MOVBU	(R1), R21
+	CMP	R20, R21
 	BNE	not_equal
 equal:
 	MOVD	$1, R0
